### ISSUE TITLE
[BUG] available_tickers function return contradicts get_quote function return

### ISSUE THREAD

**Issue by Opt33** [serkor1/cryptoQuotes#37]
**Session Information**
```r
R version 4.3.2 (2023-10-31 ucrt)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19045)

Matrix products: default


locale:
[1] LC_COLLATE=English_United States.utf8  LC_CTYPE=English_United States.utf8    LC_MONETARY=English_United States.utf8 LC_NUMERIC=C                          
[5] LC_TIME=English_United States.utf8    

time zone: America/Phoenix
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices datasets  utils     methods   base     

other attached packages:
 [1] usethis_3.0.0      janitor_2.2.0      data.table_1.16.2  lubridate_1.9.3    forcats_1.0.0      stringr_1.5.1      dplyr_1.1.4        purrr_1.0.2       
 [9] readr_2.1.5        tidyr_1.3.1        tibble_3.2.1       ggplot2_3.5.1      tidyverse_2.0.0    cryptotrackr_1.3.3 cryptoQuotes_1.3.1 Rcpp_1.0.13       
[17] pacman_0.5.1      

loaded via a namespace (and not attached):
 [1] gtable_0.3.6      xfun_0.48         remotes_2.5.0     htmlwidgets_1.6.4 lattice_0.21-9    tzdb_0.4.0        vctrs_0.6.5       tools_4.3.2      
 [9] generics_0.1.3    curl_5.2.3        fansi_1.0.6       xts_0.14.1        pkgconfig_2.0.3   lifecycle_1.0.4   compiler_4.3.2    munsell_0.5.1    
[17] snakecase_0.11.1  httpuv_1.6.15     htmltools_0.5.8.1 yaml_2.3.10       lazyeval_0.2.2    plotly_4.10.4     urlchecker_1.0.1  later_1.3.2      
[25] pillar_1.9.0      ellipsis_0.3.2    cachem_1.1.0      sessioninfo_1.2.2 mime_0.12         tidyselect_1.2.1  digest_0.6.37     stringi_1.8.4    
[33] fastmap_1.2.0     grid_4.3.2        colorspace_2.1-1  cli_3.6.3         magrittr_2.0.3    pkgbuild_1.4.5    utf8_1.2.4        withr_3.0.2      
[41] promises_1.3.0    scales_1.3.0      timechange_0.3.0  TTR_0.24.4        httr_1.4.7        zoo_1.8-12        hms_1.1.3         memoise_2.0.1    
[49] shiny_1.9.1       knitr_1.48        miniUI_0.1.1.1    viridisLite_0.4.2 profvis_0.4.0     rlang_1.1.4       xtable_1.8-4      glue_1.8.0       
[57] renv_1.0.3        pkgload_1.4.0     rstudioapi_0.17.1 jsonlite_1.8.9    R6_2.5.1          fs_1.6.4 
```

**To Reproduce**
```r
# Empty Variable
tickers <- list()

# Kraken
tickers$kraken <- cryptoQuotes::available_tickers(source="kraken", futures=FALSE)
tickers$kraken <- tibble(ticker_id = tickers$kraken, exchange = "kraken")

btcusd_test <- cryptoQuotes::get_quote(source="kraken", 
                                    ticker="BTCUSD", 
                                    futures=FALSE, 
                                    interval="1d", 
                                    from=NULL, to=NULL)

```

**Expected behavior**
I did a sanity check to see if Kraken contained the trading pair "BTCUSD" using the "available_tickers" and "get_quote" functions. I was surprised to see that available_tickers function didn't report back "BTCUSD" as it's a basic trading pair from Kraken. I tried the get_quote function with the BTCUSD pair and it did return back data. I may be mistaken, but this seems like a contradiction.

**Comment by serkor1** [comment]
Hi @Opt33,

Thank you for using {cryptoQuotes}. This is quite interesting - I can replicate the behavior if I use `ETHUSD` or `ATOMUSD`, too. I took a quick glance at the at the tradable asset pair endpoint from directly from the exchange - and there is no obvious reason for this behavior that stands out at first glance.

On an honest side note Kraken is not easy work with as it, for example, returns empty arrays instead of throwing errors even if the API call is blatantly  wrong. So it wouldn't surprise my that "something else" happens on their side.

I will investigate this issue and see if it is something that can be fix on my side, or if there at least is an explanation of this behavior.

Thank you for taking the time doing the sanity check, and posting the issue. :muscle:

**Comment by serkor1** [comment]
I realized that I didn't quite answer your issue. {cryptoQuotes} is working downstream, and (mostly) returns the queried data as is - in cases where data is wrangled all values are being thoroughly tested. Currently there are around 1000 unit tests to ensure that that the data is not corrupted in any case or form.

When it comes to what the `available_tickers()`-function returns there is only subsetting on the client side. Given this I am not sure if this issue even classifies as traditional bug. I would call it an unexpected behavior, if anything. The `BTCUSD` thing can't be explained with direct API queries, or the API documentation.

This is what is going to happen on my side from here,

- [ ] File a ticket to Kraken developer portal regarding this inconsistency
- [ ] Consider using altnames instead of array names for tickers as these appears more understandable in layman's terms. This is not a braking change as the API (it seems) reacts equally to these

This issue is interesting - and finding a decent solution to it such that the functions behaves in a predictable way will improve the package by a lot; so, again, thank you so much for spending the time on reporting this issue. This is much appreciated :pray:

Did I miss anything? And is there anything else I *should* do in this regard?

Best,

**Comment by Opt33** [comment]
Hi @serkor1 ,
Thanks for responding so quickly, but more importantly thank you for the effort directed at this R package! I like the concept of making (difficult to obtain) data more accessible. 🙏

Regarding your response, I think you covered all the bases and then some. My only concern is if other crypto exchange API's are also producing the same unexpected behavior?

As far as other suggestions to mitigate this unexpected behavior, I'd have to think about it. Should I open ticket(s) on GitHub for each?

**Comment by serkor1** [comment]
Thank you for the recognition - and I completely agree, data should be easily accessible and free. 💪 
- It's not entirely unthinkable that this behavior is present across exchanges. In the development version I have added crypto.com and HTX. If anything I'd expect these and Bitmart to misbehave 🤣 

If you do encounter more inconsistencies like this, please continue on this issue - then I'll work on it. 

Now that we have moved away from daylight saving time new bugs have been introduced in some of the source code, which I didn't anticipate when developing this - I expect to fix these bugs, and whatever you find during the weekend and publish an upgrade on Sunday if all goes well. 😄

**Comment by serkor1** [comment]
Hi @Opt33,

Here is what Kraken support wrote (If anyone from Kraken is reading along - much love to you ❤️ ):

> You are correct. When you query the pair value as "BTCUSD," it should return details about the market/pair. The same applies to "btcusd," "xbtusd," "xxbtzusd," and their uppercase versions as well. This is an intended feature to support backward compatibility. Initially, Kraken used "XBT" as the symbol/ticker for Bitcoin, but as of April 26, 2021, it was changed to "BTC" for most purposes. "XBT" is still used in the API, Futures API and log downloads, OTC desk, and History Exports.

This means that even if I wanted to fix it, I can't. It is all about backwards compatibility. If I wanted I could hardcode it, but that introduces a whole new vector of possible bugs.

I will close this issue - if you have anything else feel free to reopen the issue or post a new one. Thank you! 🙏 

Best,

@serkor1

**Comment by Opt33** [comment]
Hey @serkor1 , 

What I'm gathering from Kraken's response is in order to maintain backward compatibility they will return a list of all available (non-duplicate) cryptocurrency identifiers. However, when querying all available cryptocurrency identifiers, there's a high probability of returning duplicate cryptocurrency data?

So I went (a little) down the rabbit hole and determined that in the past Kraken may have attempted to follow International Standards Organization (ISO) implementations for financial systems:

ISO 20022
News: https://www.binance.com/en/square/post/5269189407018
Note: There seems to be a handful of cryptocurrencies that are compliant with this standard and growing.

ISO 4217
Note: The reason XBT has an X in the front is because Bitcoin doesn't have a country of origin.

ISO 24165
News: https://www.iso.org/news/ref2729.html
Standard: https://www.iso.org/standard/80601.html

So who knows? Maybe in the future ISO will have finalized their standards and there will be a lookup table of cryptocurrency identifiers :)

Thanks following up with Kraken about this!
@Opt33

**Comment by serkor1** [comment]
I love these kinds of rabbit-holes! If there were a common standard it would make wonders on so many levels, honestly. However, as it is now it also makes for an interesting application that I would love to take a crack at, at some point. See for example this wiki-entry: [https://github.com/serkor1/cryptoQuotes/wiki/Tickers#ticker-naming-convention](https://github.com/serkor1/cryptoQuotes/wiki/Tickers#ticker-naming-convention).

I highly doubt there will be any duplicate data when querying all identifiers from Kraken; their API, especially their futures API, is not easy to work with, but it's solid. 

The backwards compatibility, to my understanding, implies that coin A will be returned regardless of whether you query it using a, aa or aaa. But the API will return only the latest set identifier when querying all available identifiers. So they are working like an alias, like it is currently. Like in this specific case of how BTCUSD returns values for XBTUSD.

Exciting times!
